In [5]:
import joblib
import pandas as pd
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
warnings.filterwarnings('ignore', category=DeprecationWarning)

tr_features = pd.read_csv('C:/Users/Admin/Springboard ML/train_features.csv')
tr_labels = pd.read_csv('C:/Users/Admin/Springboard ML/train_labels.csv', header=None)

In [11]:
# we're looking for the proper kernel to make our data linearly seperable
# this is called using the kernel trick!

def print_results(results):
    print('BEST PARAMS: {}\n'.format(results.best_params_))
    
    means=results.cv_results_['mean_test_score']
    stds=results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean,3), round(std*2.3),params))

In [12]:
# create svc object
# we could hard code hyperparameters, but we will use gridsearch instead
svc = SVC()
parameters = {
    'kernel': ['linear','rbf'], # testing out these two kernels
    'C':[0.1, 1, 10] # it can be a little slow, so using smaller counts
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(tr_features, tr_labels.values.ravel()) # .ravel changing labels from column vector to array
# this will use hyperparameters in kernel and C as a grid and try to find the best combo
print_results(cv)

BEST PARAMS: {'C': 0.1, 'kernel': 'linear'}

0.796 (+/-0) for {'C': 0.1, 'kernel': 'linear'}
0.654 (+/-0) for {'C': 0.1, 'kernel': 'rbf'}
0.796 (+/-0) for {'C': 1, 'kernel': 'linear'}
0.661 (+/-0) for {'C': 1, 'kernel': 'rbf'}
0.796 (+/-0) for {'C': 10, 'kernel': 'linear'}
0.684 (+/-0) for {'C': 10, 'kernel': 'rbf'}


linear kernel is clearly doing better than rbf
also, it seems that there are three values that are almost* the same, but the gridsearch chose the best as c=0.1, linear at 0.796


In [14]:
cv.best_estimator_
# it could be useful to pickle this to be used in other notebooks or for other analysis!


SVC(C=0.1, kernel='linear')